### PR TITLE
SONARJAVA-2087: Rewrite recursion imperatively

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 1]
```diff
@@ -94,22 +96,39 @@ private FlowComputation fork(SymbolicValue symbolicValue) {
...
+        this.node = node;
+        this.trackSymbol = trackSymbol;
       }
     }
-    for (ExplodedGraph.Node parent : currentNode.getParents()) {
-      run(parent, newTrackSymbol);
+
+    Deque<NodeSymbol> workList = new ArrayDeque<>();
+    workList.add(new NodeSymbol(aNode, aSymbol));
+    while (!workList.isEmpty()) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS135)

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 2]
```diff
@@ -94,22 +96,39 @@ private FlowComputation fork(SymbolicValue symbolicValue) {
...
-    visited.add(currentNode);
-
-    Symbol newTrackSymbol = trackSymbol;
-    if (currentNode.programPoint.syntaxTree() != null) {
-      newTrackSymbol = addFlowFromLearnedSymbols(currentNode, trackSymbol);
-      Stream<Constraint> learnedConstraints = addFlowFromLearnedConstraints(currentNode);
-      if (learnedConstraints.anyMatch(terminateTraversal)) {
-        return;
+  private void run(@Nullable final ExplodedGraph.Node aNode, @Nullable final Symbol aSymbol) {
+    class NodeSymbol {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Make this a "static" inner class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2694)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [FlowComputation.java#L165](https://github.com/SonarSource/sonar-java/blob/0778c12713b6c4de5be3757f3b43804e8d9337c0/java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java#L165): Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [FlowComputation.java#L169](https://github.com/SonarSource/sonar-java/blob/0778c12713b6c4de5be3757f3b43804e8d9337c0/java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java#L169): Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java (hunk 2 above)
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: Why a local class and not a static inner class ?

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 3]
```diff
@@ -94,22 +96,39 @@ private FlowComputation fork(SymbolicValue symbolicValue) {
...
+      ExplodedGraph.Node node;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: If I'm correct those fields should be final.

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 4]
```diff
@@ -94,22 +96,39 @@ private FlowComputation fork(SymbolicValue symbolicValue) {
...
-    }
-    visited.add(currentNode);
-
-    Symbol newTrackSymbol = trackSymbol;
-    if (currentNode.programPoint.syntaxTree() != null) {
-      newTrackSymbol = addFlowFromLearnedSymbols(currentNode, trackSymbol);
-      Stream<Constraint> learnedConstraints = addFlowFromLearnedConstraints(currentNode);
-      if (learnedConstraints.anyMatch(terminateTraversal)) {
-        return;
+  private void run(@Nullable final ExplodedGraph.Node aNode, @Nullable final Symbol aSymbol) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: Why renaming the parameters here ? Not sure this is actually clearer.

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 5]
```diff
@@ -94,22 +96,39 @@ private FlowComputation fork(SymbolicValue symbolicValue) {
...
     }
-    for (ExplodedGraph.Node parent : currentNode.getParents()) {
-      run(parent, newTrackSymbol);
+
+    Deque<NodeSymbol> workList = new ArrayDeque<>();
+    workList.add(new NodeSymbol(aNode, aSymbol));
+    while (!workList.isEmpty()) {
+      NodeSymbol nodeSymbol = workList.pop();
+      ExplodedGraph.Node node = nodeSymbol.node;
+      Symbol symbol = nodeSymbol.trackSymbol;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: This variable is actually not required and its initializer can be inlined into its first usage.

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 6]
```diff
@@ -94,22 +96,38 @@ private FlowComputation fork(SymbolicValue symbolicValue) {
...
-      newTrackSymbol = addFlowFromLearnedSymbols(currentNode, trackSymbol);
-      Stream<Constraint> learnedConstraints = addFlowFromLearnedConstraints(currentNode);
-      if (learnedConstraints.anyMatch(terminateTraversal)) {
-        return;
+  }
+
+  private void run(@Nullable final ExplodedGraph.Node node, @Nullable final Symbol trackSymbol) {
+    Deque<NodeSymbol> workList = new ArrayDeque<>();
+    workList.add(new NodeSymbol(node, trackSymbol));
+    while (!workList.isEmpty()) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: MAJOR Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [FlowComputation.java#L164](https://github.com/SonarSource/sonar-java/blob/e87f9980ce521e61902d9b2e1ba6dc688443a3dd/java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java#L164): Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [FlowComputation.java#L168](https://github.com/SonarSource/sonar-java/blob/e87f9980ce521e61902d9b2e1ba6dc688443a3dd/java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java#L168): Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java (hunk 4 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: I renamed the argument to keep the name inside the method same as before

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [FlowComputation.java#L164](https://github.com/SonarSource/sonar-java/blob/effc78e31a88e62357f07108907f5c94fa8a29f0/java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java#L164): Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)
1. ![MAJOR][MAJOR] [FlowComputation.java#L168](https://github.com/SonarSource/sonar-java/blob/effc78e31a88e62357f07108907f5c94fa8a29f0/java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java#L168): Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**